mm2src/coins/lp_coins.rs

```diff
@@ -461,43 +486,218 @@ pub enum TradePreimageValue {
+pub enum WithdrawError {
+    #[display(
+        fmt = "Not enough {} to withdraw: available {}, required at least {}",
+        coin,
+        available,
+        required
+    )]
+    NotSufficientBalance {
+        coin: String,
+        available: BigDecimal,
+        required: BigDecimal,
+    },
+    #[display(fmt = "Balance is zero")]
+    ZeroBalanceToWithdrawMax,
+    #[display(fmt = "The amount {} is too small", amount)]
+    AmountIsTooSmall { amount: BigDecimal },
+    #[display(fmt = "Invalid address: {}", _0)]
+    InvalidAddress(String),
+    #[display(fmt = "Invalid fee policy: {}", _0)]
+    InvalidFeePolicy(String),
+    #[display(fmt = "No such coin {}", coin)]
+    NoSuchCoin { coin: String },
+    #[display(fmt = "Transport error: {}", _0)]
+    Transport(String),
+    #[display(fmt = "Internal error: {}", _0)]
+    InternalError(String),
+}
+
+impl HttpStatusCode for WithdrawError {
+    fn status_code(&self) -> StatusCode {
         match self {
-            TradePreimageError::NotSufficientBalance(e) => write!(f, "Not sufficient balance: {}", e),
-            TradePreimageError::Other(e) => write!(f, "{}", e),
+            WithdrawError::NotSufficientBalance { .. }
+            | WithdrawError::ZeroBalanceToWithdrawMax
+            | WithdrawError::AmountIsTooSmall { .. }
+            | WithdrawError::InvalidAddress(_)
+            | WithdrawError::InvalidFeePolicy(_)
+            | WithdrawError::NoSuchCoin { .. } => StatusCode::BAD_REQUEST,
+            WithdrawError::Transport(_) | WithdrawError::InternalError(_) => StatusCode::INTERNAL_SERVER_ERROR,
         }
     }
 }
 
-impl Traceable for TradePreimageError {
-    fn trace(self, source: TraceSource) -> Self {
-        match self {
-            TradePreimageError::NotSufficientBalance(e) => {
-                TradePreimageError::NotSufficientBalance(source.with_msg(&e))
-            },
-            TradePreimageError::Other(e) => TradePreimageError::Other(source.with_msg(&e)),
+impl From<NumConversError> for WithdrawError {
+    fn from(e: NumConversError) -> Self { WithdrawError::InternalError(e.to_string()) }
+}
+
+impl From<BalanceError> for WithdrawError {
+    fn from(e: BalanceError) -> Self {
+        match e {
+            BalanceError::Transport(error) | BalanceError::InvalidResponse(error) => WithdrawError::Transport(error),
+            BalanceError::Internal(internal) => WithdrawError::InternalError(internal),
         }
     }
 }
 
-impl From<GenerateTransactionError> for TradePreimageError {
-    fn from(e: GenerateTransactionError) -> Self {
+impl From<CoinFindError> for WithdrawError {
+    fn from(e: CoinFindError) -> Self {
         match e {
-            GenerateTransactionError::EmptyUtxoSet => {
-                TradePreimageError::NotSufficientBalance(GenerateTransactionError::EmptyUtxoSet.to_string())
+            CoinFindError::NoCoinsContext(internal) => WithdrawError::InternalError(internal),
```

Comment:
I think we can simply `unwrap/expect` instead. In the future, I would like to refactor the MmCtx to ensure that every required context is available when required.

Reply:
Would be nice. I'll do it!

Reply:
Done